Nodejs/Product/Nodejs/Intellisense/ExpansionClient.cs

```diff
@@ -0,0 +1,332 @@
+                    // need to insert a pass statement.  This is the purpose of the "SurroundsWithStatement"
+                    // snippet type.
+                    //
+                    // But, to make things even more complicated, we don't have a good indication of what's the 
+                    // template text vs. what's the selected text.  We do have access to the original template,
+                    // but all of the values have been replaced with their default values when we get called
+                    // here.  So we need to go back and re-apply the template, except for the $selected$ part.
+                    //
+                    // Also, the text has \n, but the inserted text has been replaced with the appropriate newline
+                    // character for the buffer.
+                    var templateText = codeNode.text.Replace("\n", _textView.Options.GetNewLineCharacter());
+                    foreach (var decl in declList) {
+                        string defaultValue;
+                        if (ErrorHandler.Succeeded(_session.GetFieldValue(decl, out defaultValue))) {
+                            templateText = templateText.Replace("$" + decl + "$", defaultValue);
+                        }
+                    }
+                    templateText = templateText.Replace("$end$", "");
+
+                    // we can finally figure out where the selected text began witin the original template...
+                    int selectedIndex = templateText.IndexOf("$selected$");
+                    if (selectedIndex != -1) {
+                        var selection = _textView.Selection;
+
+                        // now we need to get the indentation of the $selected$ element within the template,
+                        // as we'll need to indent the selected code to that level.
+                        string indentation = GetTemplateSelectionIndentation(templateText, selectedIndex);
+
+                        var start = _selectionStart.GetPosition(_textView.TextBuffer.CurrentSnapshot);
+                        var end = _selectionEnd.GetPosition(_textView.TextBuffer.CurrentSnapshot);
+                        if (end < start) {
+                            // we didn't actually have a selction, and our negative tracking pushed us
+                            // back to the start of the buffer...
+                            end = start;
+                        }
+                        var selectedSpan = Span.FromBounds(start, end);
+
+                        if (surroundsWithStatement &&
+                            String.IsNullOrWhiteSpace(_textView.TextBuffer.CurrentSnapshot.GetText(selectedSpan))) {
+                            // we require a statement here and the user hasn't selected any code to surround,
+                            // so we insert a pass statement (and we'll select it after the completion is done)

```

Comment:
Luckily JavaScript is less annoying than python, so this can all be simplified :smiley:

The reason we did this in python was because python requires that code blocks cannot be empty. So the following code isn't syntactically valid. 
```python
while True:
    # nothingness
``` 

Instead, python has a no-op ("do nothing") statement for these cases.
```python
while True:
    pass
````